compiler/src/lnntoamm/Fn.ts

```diff
@@ -126,76 +120,64 @@ export default class Fn {
     );
   }
 
-  getType(): FunctionType {
-    return this.fnType;
+  acceptsTypes(tys: Type[]): boolean {
+    let params = Object.values(this.params);
+    if (params.length !== tys.length) {
+      return false;
+    }
+    for (let ii = 0; ii < params.length; ii++) {
+      if (!params[ii].ty.compatibleWithConstraint(tys[ii])) {
+        return false;
+      }
+    }
+    return true;
   }
 
   asHandler(amm: Output, event: string) {
-    let handlerArgs = [];
-    for (let arg of Object.keys(this.args)) {
-      handlerArgs.push([arg, this.args[arg].ty.breakdown()]);
+    let handlerParams = [];
+    for (let param of Object.keys(this.params)) {
+      handlerParams.push([param, this.params[param].ty.breakdown()]);
     }
-    amm.addHandler(event, handlerArgs, this.retTy.breakdown());
+    amm.addHandler(event, handlerParams, this.retTy.breakdown());
     let isReturned = false;
     for (let ii = 0; ii < this.body.length; ii++) {
       const stmt = this.body[ii];
-      if (stmt instanceof Dec || stmt instanceof Assign || stmt instanceof Emit) {
-        stmt.inline(amm);
-      } else if (stmt instanceof Exit) {
-        stmt.inline(amm);
+      stmt.inline(amm);
+      if (stmt instanceof Exit) {
         isReturned = true;
         if (ii !== this.body.length - 1) {
           throw new Error(`hmmmm... unreachable statements probably should've been caught earlier?`);
         }
-        break;
-      } else {
-        throw new Error(`did not expect to inline stmt: ${stmt}`);
       }
     }
     if (!isReturned) {
-      if (this.retTy.breakdown() !== opcodes().get('void')) {
+      if (!this.retTy.compatibleWithConstraint(opcodes().get('void'))) {
         throw new Error(`no return value for function`);
       }
       amm.exit();
     }
   }
 
-  // TODO: this will have to change in order to call fns multiple times - maybe deep cloning?
-  inline(amm: Output, args: Ref[], assign: string, kind: 'const' | 'let' | '') {
-    let argNames = Object.keys(this.args);
-    if (argNames.length !== args.length) {
-      // this should be caught by Call, it's just a sanity check
-      throw new Error(`number of arguments off`);
+  inline(amm: Output, args: Ref[], kind: AssignKind, name: string, ty: Builtin) {
+    let paramDefs = Object.values(this.params);
+    if (args.length !== paramDefs.length) {
+      throw new Error(`function call argument mismatch`);
```

Comment:
This shouldn't ever happen in reality, correct?

Reply:
yeah, I just check many times to make absolutely sure. This error should never be thrown, since `Call` filters out functions that have an argument mismatch